nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQueryStreaming.java

```diff
@@ -0,0 +1,206 @@
+                        request.addRow(convertMapRecord(currentRecord.toMap()));
+                        nbrecord++;
+                    }
+                }
+            }
+
+            request.setIgnoreUnknownValues(context.getProperty(IGNORE_UNKNOWN).asBoolean());
+            request.setSkipInvalidRows(context.getProperty(SKIP_INVALID_ROWS).asBoolean());
+
+            InsertAllResponse response = getCloudService().insertAll(request.build());
+
+            final Map<String, String> attributes = new HashMap<>();
+
+            if (response.hasErrors()) {
+                getLogger().log(LogLevel.WARN, "Failed to insert {} of {} records into BigQuery {} table.", new Object[] { response.getInsertErrors().size(), nbrecord, tableName });
+                if (getLogger().isDebugEnabled()) {
+                    for (long index : response.getInsertErrors().keySet()) {
+                        for (BigQueryError e : response.getInsertErrors().get(index)) {
+                            getLogger().log(LogLevel.DEBUG, "Failed to insert record #{}: {}", new Object[] { index, e.getMessage() });
+                        }
+                    }
+                }
+
+                attributes.put(BigQueryAttributes.JOB_NB_RECORDS_ATTR, Long.toString(nbrecord - response.getInsertErrors().size()));
+
+                flowFile = session.penalize(flowFile);
+                flowFile = session.putAllAttributes(flowFile, attributes);
+                session.transfer(flowFile, REL_FAILURE);
+            } else {
+                attributes.put(BigQueryAttributes.JOB_NB_RECORDS_ATTR, Long.toString(nbrecord));
+                flowFile = session.putAllAttributes(flowFile, attributes);
+                session.transfer(flowFile, REL_SUCCESS);
+            }
+
+        } catch (Exception ex) {
+            getLogger().log(LogLevel.ERROR, ex.getMessage(), ex);
+            flowFile = session.penalize(flowFile);
+            session.transfer(flowFile, REL_FAILURE);
+        }
+    }
+
+    private Map<String, Object> convertMapRecord(Map<String, Object> map) {
+        Map<String, Object> result = new HashMap<String, Object>();
+        for (String key : map.keySet()) {
+            Object obj = map.get(key);
+            if (obj instanceof MapRecord) {
+                result.put(key, convertMapRecord(((MapRecord) obj).toMap()));
+            } else if (obj instanceof Object[]) {
+                if (((Object[]) obj).length > 0) {
+                    if (((Object[]) obj)[0] instanceof MapRecord) {
+                        List<Map<String, Object>> lmapr = new ArrayList<Map<String, Object>>();
+                        for (Object mapr : ((Object[]) obj)) {
+                            lmapr.add(convertMapRecord(((MapRecord) mapr).toMap()));
+                        }
+                        result.put(key, lmapr);
+                    } else {
+                        result.put(key, obj);
```

Comment:
Instead of doing this in multiple branches I think it would be a bit nicer to indicate the success special cases (in a boolean for eg) and do this at the end of the body of the outer loop.